src/dict.c

```diff
@@ -1421,24 +1420,31 @@ static int _dictExpandIfNeeded(dict *d)
         (dict_can_resize != DICT_RESIZE_FORBID &&
          d->ht_used[0] / DICTHT_SIZE(d->ht_size_exp[0]) > dict_force_resize_ratio))
     {
-        if (!dictTypeReHashAllowed(d))
+        if (!dictTypeResizeAllowed(d))
             return DICT_OK;
         return dictExpand(d, d->ht_used[0] + 1);
     }
     return DICT_OK;
 }
 
-static int _dictShrinkIfNeeded(dict *d) {
+static int _dictShrinkIfNeeded(dict *d) 
+{
     /* Incremental rehashing already in progress. Return. */
     if (dictIsRehashing(d)) return DICT_OK;
     
-    if (DICTHT_SIZE(d->ht_size_exp[0]) > DICT_HT_INITIAL_SIZE && 
-        ((dict_can_resize == DICT_RESIZE_ENABLE &&  
-          d->ht_used[0] * 100 / DICTHT_SIZE(d->ht_size_exp[0]) < HASHTABLE_MIN_FILL) ||
-         (dict_can_resize != DICT_RESIZE_FORBID &&
-          d->ht_used[0] * 100 / DICTHT_SIZE(d->ht_size_exp[0]) < dict_force_shrink_ratio))) 
+    /* If the size of hash table is DICT_HT_INITIAL_SIZE, don't shrink it. */
+    if (DICTHT_SIZE(d->ht_size_exp[0]) == DICT_HT_INITIAL_SIZE) return DICT_OK;
+
+    /* If we reached the 1:10 ratio, and we are allowed to resize the hash
+     * table (global setting) or we should avoid it but the ratio between
+     * elements/buckets is below the lower threshold, we shrink to
+     * the next power of 2 greater than elements. */
```

Comment:
```suggestion
    /* If we reached below 1:10 elements/buckets ratio, and we are allowed to resize
     * the hash table (global setting) or we should avoid it but the ratio is below 1:50,
     * we'l trigger a resize of the hash table. */
```

Reply:
```
/* Trigger a resize of the hash table if either of these conditions is met:
     * 1. The elements-to-buckets ratio falls below 1:10, and resizing the hash table is permitted by a global setting.
     * 2. The elements-to-buckets ratio falls below 1:50, even if resizing is generally to be avoided.
*/
```

Powered by ChatGPT 😅

Reply:
little too lengthy to my taste, but whatever..